segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1350 @@
+                }
+
+                // Update the read index.
+                val readIndex = getReadIndex(streamSegmentName);
+                for (val entry : newReadIndexEntries.entrySet()) {
+                    readIndex.put(entry.getKey(), entry.getValue());
+                }
+
+                Duration elapsed = timer.getElapsed();
+                log.debug("write - segment={} offset={} length={} latency={}.", handle.getSegmentName(), offset, length, elapsed.toMillis());
+                LoggerHelpers.traceLeave(log, "write", traceId, handle, offset);
+                return null;
+            } catch (StorageMetadataVersionMismatchException ex) {
+                throw ex;
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            } catch (Exception ex) {
+                throw ex;
+            } finally {
+                if (!isCommited && chunksAddedCount > 0) {
+                    // Collect garbage.
+                    collectGarbage(newReadIndexEntries.values());
+                }
+            }
+        });
+    }
+
+    /**
+     * Gets whether given segment is a critical storage system segment.
+     * @param streamSegmentName Name of the segment.
+     * @return True if this is a storage system segment.
+     */
+    private boolean isStorageSystemSegment(String streamSegmentName) {
+        return null != systemJournal && systemJournal.isStorageSystemSegment(streamSegmentName);
+    }
+
+    /**
+     * Adds a system log.
+     * @param systemLogRecords
+     * @param streamSegmentName Name of the segment.
+     * @param offset Offset at which new chunk was added.
+     * @param oldChunkName Name of the previous last chunk.
+     * @param newChunkName Name of the new last chunk.
+     */
+    private void addSystemLogRecord(ArrayList<String> systemLogRecords, String streamSegmentName, long offset, String oldChunkName, String newChunkName) {
+        val systemLogRecord = systemJournal.getChunkAddedRecord(streamSegmentName,
+            offset,
+            oldChunkName == null ? null : oldChunkName,
+            newChunkName);
+        systemLogRecords.add(systemLogRecord);
+    }
+
+    /**
+     * Delete the garbage chunks.
+     * @param chunksTodelete List of chunks to delete.
+     */
+    private void collectGarbage(Collection<String> chunksTodelete) {
+        for (val chunkTodelete : chunksTodelete) {
+            try {
+                if (chunkStorage.exists(chunkTodelete)) {
+                    chunkStorage.delete(chunkStorage.openWrite(chunkTodelete));
+                    log.debug("collectGarbage - chunk={}.", chunkTodelete);
+                }
+            } catch (FileNotFoundException e) {
+                log.debug("Could not delete garbage chunk {}", chunkTodelete);
+            } catch (Exception e) {
+                log.warn("Could not delete garbage chunk {}", chunkTodelete);
```

Comment:
what happens in this case ?
are we leaving garbage forever
or do we have some kind of garbage collector that sometimes looks for orphan chunks ?

Reply:
My plan is to implement this in phases
0. Initial implementation - Best effort, try to delete but leave garbage if you have to. 
1. This iteration - Add names of deleted chunks to an in-memory queue and periodically delete chunks.
2. Next iteration - Mark chunks deleted in metadata and run Background garbage collection job that scans metadata.